### DISCUSSION TITLE
Large Tables

### DISCUSSION THREAD

**Discussion by waywardcoder** [weldsorm/welds#35] [General]
First i want to say how much i appreciate the work that goes into welds, i daily drive it on several projects.  I am using mssql.  I have a contact table that has about 348K entries in it.  I need to load those and process them.  When i execute the query in ssms it takes about 4 seconds to execute.  When i try using welds for this, it never returns (or at least i have not waited long enough, longest wait was 15m).  My code is extremely simple:
   
  SQL : SELECT distinct     [Name]     ,[address1]     ,[address2]     ,[city]     ,[state]     ,[zip] FROM [Contacts]

    pub async fn get_mortgagees(&mut self) -> Vec<DbState<Contact>> {
        let args = welds::query::clause::ParamArgs::new();
        let query = Contact::from_raw_sql(&CONTACT_SQL, &args, &self.client).await;
        query.unwrap_or_default()
    }

I have also tried :

    pub async fn get_contacts(&mut self) -> Vec<DbState<Contact>> {
       Contact::all().run(&self.client).await.unwrap()
    }

I also tried this just to try and see what is happening and it seems to quit at about 4K rows:

 pub async fn get_mortgagees(&mut self) -> Vec<DbState<Contact>> {
        let mut contacts: Vec<DbState<Mortgagee>> = Vec::with_capacity(348000);
        let count = Contact::all().count(&self.client).await.unwrap();
        let mut processed:u64 = 0;
        event!(
            Level::INFO,
            "get_contacts"
        );
        while processed < count{
            let mut chunk = Contact::all().offset(processed as i64).limit(1000i64);
            let sql = chunk.to_sql(welds::connections::Syntax::Mssql);
            contacts.append(&mut chunk.run(&self.client).await.unwrap());
            processed += 1000;
            event!(
                Level::INFO,
                with_processed = processed,
                "get_contacts"
            );
        }
        contacts
    }

 

 what is the recommended way to do this using welds with reasonable performance?

**Comment by lex148** [comment]
Hey @waywardcoder I am a little surprised by this one. Welds should easily be able to handle 348K records at a time. I wrote a little sample project to test and was pulling back 348K rows in about 4 seconds (see below). That being said I do have a couple Ideas what could be going on. 

1) could it be something after the fetching of the records. You are getting a LOT of records. if you are trying to edit them or re-import them, you should know single row operations are very slow compared to `SELECT`. If this is the case bulk operations could really help. 

- you could try putting a `println!` right before and after the run to verify the select is the culprit. 


2) There might be other hidden SQL going on here 

- try enabling something like `log_env` with `export RUST_LOG=debug`


3) SSMS cheats. it doesn't really return all the rows it streams them in from the DB. Could you do something similar like page the results and work with only a chunk at a time? I know you shouldn't need to with 348K but when your talking about larger datasets It really can make things run faster. (side note: I have been thinking of adding a streaming row in from DB feature to welds)


4) could It be the connection to the DB (I.E. remote server?) If you are dealing with a remote server the transfer speed might be killing you.


I would love to see what you get from the sample code below. I am sure you might need to tweak it for your DB. Please keep me in the loop here. If welds is the issue that is a problem and I want to get it fixed ASAP.



Cargo.toml
```
[package]
name = "example"
version = "0.1.0"
edition = "2021"

[dependencies]
fake = "2.9.2"
rand = "0.8.5"
tokio = { version = "1", features = ["full"] }
welds = { version = "0.4.1", features = ["mssql"] }
```

main.rs
``` 
use fake::faker::address::en::*;
use fake::faker::name::en::*;
use fake::Fake;
use rand::Rng;
use welds::connections::{Client, TransactStart};
use welds::prelude::*;

#[derive(WeldsModel, Debug, Clone)]
#[welds(schema = "welds", table = "Contacts")]
struct Contact {
    name: String,
    address1: Option<String>,
    address2: Option<String>,
    city: Option<String>,
    state: Option<String>,
    zip: Option<String>,
}

fn generate_random_contact() -> Contact {
    let mut rng = rand::thread_rng();

    // Generate fake data
    let name: String = Name().fake();
    let address1: String = StreetName().fake();
    let address2: Option<String> = if rng.gen_bool(0.5) {
        Some(SecondaryAddress().fake())
    } else {
        None
    };
    let city: String = CityName().fake();
    let state: String = StateAbbr().fake();
    let zip: String = PostCode().fake();

    Contact {
        name,
        address1: Some(address1),
        address2,
        city: Some(city),
        state: Some(state),
        zip: Some(zip),
    }
}

#[tokio::main]
async fn main() -> std::result::Result<(), Box<dyn std::error::Error>> {
    let url = "server=127.0.0.1;user id=sa;password=welds!123;TrustServerCertificate=true;";
    let conn = welds::connections::mssql::connect(url).await.unwrap();

    // for batch_num in 0..35 {
    //     let mut batch = Vec::default();
    //     for _i in 0..10000 {
    //         batch.push(generate_random_contact())
    //     }
    //     welds::query::insert::bulk_insert(&conn, &batch).await?;
    //     println!("Inserted: {}", batch_num * 10000);
    // }

    let total = Contact::all().count(&conn).await?;
    println!("Table Contains: {total}");

    let sql = "SELECT distinct name, address1, address2, city, state, zip FROM welds.Contacts";
    let args = welds::query::clause::ParamArgs::new();
    let mut all = Contact::from_raw_sql(sql, &args, &conn).await?;
    println!("loaded! {}", all.len());

    Ok(())
}
```

**Comment by waywardcoder** [comment]
I tested your code and then looked at mine again. We are doing the same thing, with one small exception.  In my larger program i am using executor::block_on() with the async method loading the contacts.  This seems to cause tiberious or welds or something to hang.  Changing all the methods to async and using await everywhere (getting ride of the executor::block_on() call) seems to resolve the issue.  I will post a short example tomorrow showing the issue when using executor and async methods with welds/tiberious. Ultimately though upon closer inspection, this is likely more user error than library error.
